# 初Keras

# reuters_mlp.py
github上の[keras examples][1]の[reuters_mlp.py][2]。  

[1]: https://github.com/keras-team/keras/tree/master/examples
[2]: https://github.com/keras-team/keras/blob/master/examples/reuters_mlp.py

In [1]:
'''Trains and evaluate a simple MLP
on the Reuters newswire topic classification task.
'''
from __future__ import print_function

import numpy as np
import keras
from keras.datasets import reuters
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation
from keras.preprocessing.text import Tokenizer

max_words = 1000
batch_size = 32
epochs = 5

print('Loading data...')
(x_train, y_train), (x_test, y_test) = reuters.load_data(num_words=max_words,
                                                         test_split=0.2)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

num_classes = np.max(y_train) + 1
print(num_classes, 'classes')

print('Vectorizing sequence data...')
tokenizer = Tokenizer(num_words=max_words)
x_train = tokenizer.sequences_to_matrix(x_train, mode='binary')
x_test = tokenizer.sequences_to_matrix(x_test, mode='binary')
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

print('Convert class vector to binary class matrix '
      '(for use with categorical_crossentropy)')
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)
print('y_train shape:', y_train.shape)
print('y_test shape:', y_test.shape)

print('Building model...')
model = Sequential()
model.add(Dense(512, input_shape=(max_words,)))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

history = model.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    verbose=1,
                    validation_split=0.1)
score = model.evaluate(x_test, y_test,
                       batch_size=batch_size, verbose=1)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Using TensorFlow backend.


Loading data...
2113536/2110848 [==============================] - 4s 2us/step
8982 train sequences
2246 test sequences
46 classes
Vectorizing sequence data...
x_train shape: (8982, 1000)
x_test shape: (2246, 1000)
Convert class vector to binary class matrix (for use with categorical_crossentropy)
y_train shape: (8982, 46)
y_test shape: (2246, 46)
Building model...
Train on 8083 samples, validate on 899 samples
Epoch 1/5
8083/8083 [==============================] - 6s 710us/step - loss: 1.4291 - acc: 0.6795 - val_loss: 1.0911 - val_acc: 0.7631
Epoch 2/5
8083/8083 [==============================] - 5s 668us/step - loss: 0.7889 - acc: 0.8188 - val_loss: 0.9421 - val_acc: 0.7887
Epoch 3/5
8083/8083 [==============================] - 5s 636us/step - loss: 0.5501 - acc: 0.8651 - val_loss: 0.8915 - val_acc: 0.7987
Epoch 4/5
8083/8083 [==============================] - 6s 804us/step - loss: 0.4152 - acc: 0.9004 - val_loss: 0.8775 - val_acc: 0.8065
Epoch 5/5
2246/2246 [========================

# reuters_mlp.py(tfidf)

In [2]:
'''Trains and evaluate a simple MLP
on the Reuters newswire topic classification task.
'''
from __future__ import print_function

import numpy as np
import keras
from keras.datasets import reuters
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation
from keras.preprocessing.text import Tokenizer

max_words = 1000
batch_size = 32
epochs = 5

print('Loading data...')
(x_train, y_train), (x_test, y_test) = reuters.load_data(num_words=max_words,
                                                         test_split=0.2)

Loading data...


In [4]:
import pickle

pickle_file="./reuters.pickle"

with open(pickle_file, 'wb') as f:
    pickle.dump(x_train, f)
    pickle.dump(y_train, f)
    pickle.dump(x_test, f)
    pickle.dump(y_test, f)

In [8]:
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')

num_classes = np.max(y_train) + 1
print(num_classes, 'classes')

print('Vectorizing sequence data...')
tokenizer = Tokenizer(num_words=max_words)
# x_train = tokenizer.sequences_to_matrix(x_train, mode='binary')
# x_test = tokenizer.sequences_to_matrix(x_test, mode='binary')
tokenizer.fit_on_sequences(x_train)
x_train = tokenizer.sequences_to_matrix(x_train, mode='tfidf')
x_test = tokenizer.sequences_to_matrix(x_test, mode='tfidf')
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

print('Convert class vector to binary class matrix '
      '(for use with categorical_crossentropy)')
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)
print('y_train shape:', y_train.shape)
print('y_test shape:', y_test.shape)

8982 train sequences
2246 test sequences
46 classes
Vectorizing sequence data...
x_train shape: (8982, 1000)
x_test shape: (2246, 1000)
Convert class vector to binary class matrix (for use with categorical_crossentropy)
y_train shape: (8982, 46)
y_test shape: (2246, 46)


In [9]:
print('Building model...')
model = Sequential()
model.add(Dense(512, input_shape=(max_words,)))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

history = model.fit(x_train, y_train,
                    batch_size=batch_size,
                    epochs=epochs,
                    verbose=1,
                    validation_split=0.1)
score = model.evaluate(x_test, y_test,
                       batch_size=batch_size, verbose=1)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Building model...
Train on 8083 samples, validate on 899 samples
Epoch 1/5
8083/8083 [==============================] - 5s 672us/step - loss: 1.3677 - acc: 0.7138 - val_loss: 0.9856 - val_acc: 0.7887
Epoch 2/5
8083/8083 [==============================] - 5s 617us/step - loss: 0.6406 - acc: 0.8567 - val_loss: 0.9046 - val_acc: 0.8065
Epoch 3/5
8083/8083 [==============================] - 6s 712us/step - loss: 0.4277 - acc: 0.8937 - val_loss: 0.9259 - val_acc: 0.8053
Epoch 4/5
8083/8083 [==============================] - 6s 706us/step - loss: 0.3265 - acc: 0.9223 - val_loss: 0.9604 - val_acc: 0.8053
Epoch 5/5
2246/2246 [==============================] - 0s 130us/step
Test score: 1.0029771964887584
Test accuracy: 0.7951914514692787


# 結果
## reuters_mlp.py(オリジナル。binary)
79%
## reuters_mlp.py(tfidf)
79%